src/Microsoft.AspNetCore.Authentication.MicrosoftAccount/MicrosoftAccountOptions.cs

```diff
@@ -30,9 +30,7 @@ public MicrosoftAccountOptions()
             ClaimResolvers.Add(ClaimTypes.Name, "displayName");
             ClaimResolvers.Add(ClaimTypes.GivenName, "givenName");
             ClaimResolvers.Add(ClaimTypes.Surname, "surname");
-            // Last one wins, if present
-            ClaimResolvers.Add(ClaimTypes.Email, "userPrincipalName");
-            ClaimResolvers.Add(ClaimTypes.Email, "mail");
+            ClaimResolvers.AddCustom(ClaimTypes.Email, user => user.Value<string>("mail") ?? user.Value<string>("userPrincipalName"));
```

Comment:
Really? I thought the new version was supper clean and easy. It sounds like you want some more explicit names though.